README.rst
VKvideo lens
============
Lens for searching video in Vkontakte and open it via totem.


Installation
============
For ubuntu user `ppa <https://launchpad.net/~nvbn-rm/+archive/ppa>`_:

``add-apt-repository ppa:nvbn-rm/ppa``

``apt-get update``

``apt-get install vkvideo`` 

And restart unity:

``unity --replace``

Developers can clone this repository and install via ``setup.py``


debian/changelog
vkvideo (9-0~precise) precise; urgency=low

  * Fix new urls
  * Fix settings and authorisation

 -- Vladimir Yakovlev <nvbn.rm@gmail.com>  Mon, 26 Nov 2012 05:24:15 +0400

vkvideo (8-0~precise) precise; urgency=low

  * Fix previews

 -- Vladimir Yakovlev <nvbn.rm@gmail.com>  Sat, 21 Oct 2012 22:32:00 +0400

vkvideo (7-0~precise) precise; urgency=low

  * Fix package
  * Fix access denied

 -- Vladimir Yakovlev <nvbn.rm@gmail.com>  Sat, 20 Sep 2012 03:44:00 +0400

vkvideo (6-0~precise) precise; urgency=low

  * Fix locks
  * Previews support

 -- Vladimir Yakovlev <nvbn.rm@gmail.com>  Sun, 23 Sep 2012 21:55:00 +0400

vkvideo (5-0~precise) precise; urgency=low

  * Static vkontakte

 -- Vladimir Yakovlev <nvbn.rm@gmail.com>  Sun, 11 May 2012 23:12:24 +0400

vkvideo (4-0~precise) precise; urgency=low

  * Full name view in vertical tile
  * Fix python-support error

 -- Vladimir Yakovlev <nvbn.rm@gmail.com>  Sun, 11 May 2012 02:35:12 +0400

vkvideo (3-0~precise) precise; urgency=low

  * Add settings

 -- Vladimir Yakovlev <nvbn.rm@gmail.com>  Sun, 1 May 2012 19:20:08 +0400

vkvideo (2-0~precise) precise; urgency=low

  * Fix packaging

 -- Vladimir Yakovlev <nvbn.rm@gmail.com>  Sun, 30 Apr 2012 01:45:19 +0400

vkvideo (1-0~precise) precise; urgency=low

  * Initial release

 -- Vladimir Yakovlev <nvbn.rm@gmail.com>  Sun, 29 Apr 2012 00:11:12 +0400


vkvideo/auth.py
from gi.repository import Gtk, WebKit, Notify
import re
import keyring
import os
import gettext


path = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'locale')
if not os.path.isdir(path):
    path = '/usr/share/locale/'
gettext.bindtextdomain('vklensauth', path)
gettext.textdomain('vklensauth')
_ = gettext.gettext


class AuthDialog(Gtk.Window):
    """Authentication dialog"""
    def __init__(self, *args, **kwargs):
        """Open vk auth page"""
        super(AuthDialog, self).__init__(*args, **kwargs)
        self.webview = WebKit.WebView()
        self.add(self.webview)
        self.webview.open("""http://oauth.vk.com/authorize?
        client_id=2739195&
        scope=video,offline&
        redirect_uri=oauth.vk.com/blank.html&
        display=page&
        response_type=token""")
        self.webview.connect('load-finished', self.finished)

    def finished(self, widget, user_data):
        """Page load fifnshed"""
        uri = widget.get_uri()
        if 'access_token' in uri:
            token = re.match('.*access_token=(.*)&expires.*', uri).groups()[0]
            keyring.set_password(
                'vkvideo-lens', 'access_token', token,
            )
            Notify.init("vkvideo")
            Notify.Notification.new(
                'vkvideo',
                 _('authorisations success'),
            '').show()
            Gtk.main_quit()


def main():
    win = AuthDialog(title=_('Authorise'))
    win.connect("delete-event", Gtk.main_quit)
    win.show_all()
    Gtk.main()


if __name__ == "__main__":
    main()


vkvideo/settings.py
from gi.repository import Gtk, Notify, GConf
import os
import gettext


path = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'locale')
if not os.path.isdir(path):
    path = '/usr/share/locale/'
gettext.bindtextdomain('vksettings', path)
gettext.textdomain('vksettings')
_ = gettext.gettext


class SettingsDialog(Gtk.Dialog):
    """Settings dialog"""
    QUALITYS = ['240', '360', '480', '720']
    DEFAULT_QUALITY = '720'
    DEFAULT_PLAYER = 'totem'
    PLAYER_KEY = '/apps/unity-vkvideo-lens/player'
    QUALITY_KEY = '/apps/unity-vkvideo-lens/quality'

    def __init__(self, *args, **kwargs):
        """Open settings dialog"""
        super(SettingsDialog, self).__init__(*args, **kwargs)
        self.settings = GConf.Client.get_default()
        self.set_default_size(150, 100)
        box = self.get_content_area()
        vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=6)
        box.add(vbox)
        hbox = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=6)
        hbox.add(Gtk.Label(_('Prefered player:')))
        self.player = Gtk.Entry()
        current_player = self.settings.get_string(SettingsDialog.PLAYER_KEY) or SettingsDialog.DEFAULT_PLAYER
        self.player.set_text(current_player)
        hbox.add(self.player)
        vbox.add(hbox)
        hbox2 = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=6)
        hbox2.add(Gtk.Label(_('Prefered quality:')))
        self.quality = Gtk.ComboBoxText()
        for q in SettingsDialog.QUALITYS:
            self.quality.append_text(q)
        current_quality = self.settings.get_string(SettingsDialog.QUALITY_KEY) or SettingsDialog.DEFAULT_QUALITY
        self.quality.set_active(SettingsDialog.QUALITYS.index(current_quality))
        hbox2.add(self.quality)
        vbox.add(hbox2)
        self.connect('response', self.on_action)

    def on_action(self, dialog, response_id):
        if response_id == Gtk.ResponseType.OK:
            self.settings.set_string(
                SettingsDialog.QUALITY_KEY, 
                SettingsDialog.QUALITYS[self.quality.get_active()],
            )
            self.settings.set_string(
                SettingsDialog.PLAYER_KEY, 
                self.player.get_text(),
            )
            Notify.init("vkvideo")
            Notify.Notification.new(
                'vkvideo',
                 _('Settings saved success'),
            '').show()
        Gtk.main_quit()


def main():
    win = SettingsDialog(_('Settings'), None, 0,
            (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL,
             Gtk.STOCK_OK, Gtk.ResponseType.OK))
    win.connect("delete-event", Gtk.main_quit)
    win.show_all()
    Gtk.main()


if __name__ == "__main__":
    main()


vkvideo/lens.py
#!/usr/bin/python
from gi.repository import GLib, GObject, Gio, Unity, GConf, Gdk
from functools import wraps, partial
import collections
import vkontakte
import keyring
import urllib
import re
import subprocess
import gettext
import os
import threading
import time


path = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'locale')
if not os.path.isdir(path):
    path = '/usr/share/locale/'
gettext.bindtextdomain('vklens', path)
gettext.textdomain('vklens')
_ = gettext.gettext


def update_on_finish(fnc):
    @wraps(fnc)
    def wrapper(self, *args, **kwargs):
        try:
            return fnc(self, *args, **kwargs)
        except AttributeError:
            return None
        finally:
            self.update_results_model()
    return wrapper


def open_uri(player, uri):
    subprocess.Popen(player.split(' ') + [
        uri,
    ])


class VideoPage(object):
    def __init__(self, uri, max_quality=720):
        data = urllib.urlopen(uri).read()
        vars = re.search('flashvars=(.*)><', data).group(1)
        vars = dict(map(lambda part: part.split('='), vars.split('&amp;')))
        url = vars['host'] + 'u' + vars['uid'] + '/' + 'videos' + '/' + vars['vtag'] + '.%s.mp4'
        qualitys = ['720', '480', '360', '240']

        urls = map(lambda quality: (quality, url % quality),
            qualitys[qualitys.index(str(max_quality)):])
        for num, (quality, url) in enumerate(urls):
            if urllib.urlopen(url).code == 200:
                results = urls[num:]
                break
        self.thumb = vars['thumb']
        self.urls = results
        self.title = urllib.unquote_plus(vars['md_title'])

    def open(self, player, quality=0):
        open_uri(player, self.urls[quality][1])


class ActionIdle(object):
    def __init__(self):
        self.open_queue = collections.deque()
        self.page_queue = collections.deque()
        GObject.timeout_add(1000, self.run)

    def open(self, uri, max_quality, player, *args):
        self.open_queue.append((
            uri, max_quality, player,
        ))

    def run(self):
        try:
            uri, max_quality, player = self.open_queue.popleft()
            if max_quality == -1:
                open_uri(player, uri)
            else:
                page = VideoPage(uri, max_quality)
                page.open(player)
        except IndexError:
            pass
        GObject.timeout_add(1000, self.run)


class Daemon(object):
    """Lens daemon"""
    def __init__(self, session_bus_connection):
        """Init lens and create scope"""
        self.session_bus_connection = session_bus_connection
        self.lens = Unity.Lens.new("/net/launchpad/unity/lens/vkvideo", "vkvideo")
        self.lens.props.search_hint = _("Start writting video name")
        self.lens.props.visible = True
        self.lens.props.search_in_global = False
        category_icon = Gio.ThemedIcon.new("vkvideo")
        icon = Gio.ThemedIcon.new("/usr/share/icons/unity-icon-theme/places/svg/group-recent.svg")
        self.lens.props.categories = [
            Unity.Category.new(_("Videos"), category_icon,
                Unity.CategoryRenderer.HORIZONTAL_TILE,
            ), Unity.Category.new(_("Settings"), category_icon,
                Unity.CategoryRenderer.VERTICAL_TILE,
            ),
        ]
        quality = Unity.RadioOptionFilter.new('quality', _('quality'), icon, True)
        quality.add_option('with_low', _('with low'), icon)
        sort = Unity.RadioOptionFilter.new('sorting', _('sorting'), icon, True)
        sort.add_option('0', _('by date'), icon)
        sort.add_option('1', _('by length'), icon)
        sort.add_option('2', _('by rel'), icon)
        count = Unity.RadioOptionFilter.new('count', _('count'), icon, True)
        count.add_option('20', '20', icon)
        count.add_option('50', '50', icon)
        count.add_option('100', '100', icon)
        count.add_option('200', '200', icon)
        self.lens.props.filters = [quality, sort, count]
        self.scope = VKScope()
        self.lens.add_local_scope(self.scope.scope)
        self.lens.export()


class VKScope(object):
    """Scope with vk videos"""
    def __init__(self):
        """Init scope and connect signals"""
        self.search_string = ''
        self.only_hd = 1
        self.sorting = 0
        self.model = None
        self.count = 20
        self._vk = None
        self.settings = GConf.Client.get_default()
        self.scope = Unity.Scope.new("/net/launchpad/unity/scope/vkvideo")
        self.scope.connect("activate_uri", self.on_uri_activated)
        self.scope.connect("filters-changed", self.on_filter_changed)
        self.scope.connect("search-changed", self.on_search_changed)
        self.scope.export()
        self.scope.connect('preview-uri', self.on_preview_uri)
        self.action_idle = ActionIdle()

    @property
    def vk(self):
        """VK api property"""
        if not self._vk:
            token = keyring.get_password('vkvideo-lens', 'access_token')
            self._vk = vkontakte.API(token=token) if token else None
        return self._vk

    @update_on_finish
    def on_search_changed(self, scope, search, search_type, cancellable):
        """Search value changed event"""
        self.search_string = search.props.search_string.strip()
        self.model = search.props.results_model

    @update_on_finish
    def on_filter_changed(self, scope):
        """Filter value changed event"""
        self.only_hd = 1 - int(bool(scope.get_filter('quality').get_active_option()))
        sorting = scope.get_filter('sorting').get_active_option()
        if sorting:
            self.sorting = sorting.props.id
        else:
            self.sorting = 0
        count = scope.get_filter('count').get_active_option()
        if count:
            self.count = count.props.id
        else:
            count = 20

    def update_results_model(self):
        """Update results"""
        model = self.model
        if model is not None:
            model.clear()
            if self.vk:
                try:
                    self._last_result = self.vk.get(
                        'video.search', q=self.search_string,
                        hd=self.only_hd, sort=self.sorting,
                        count=self.count,
                    )
                    for num, entry in enumerate(self._last_result):
                        model.append(str(num),
                            entry['thumb'], 0, "video/mp4",
                            entry['title'], "",
                        entry['player'])
                except Exception:  # Prevent apport errors
                    pass
            model.append('vkvideo',
                'vkvideo', 1, 'image/png',
                _('Authorisation'), "",
            "vkvideo")
            model.append('vksettings',
                'vkvideo', 1, 'image/png',
                _('Settings'), "",
            "vksettings")

    def on_uri_activated(self, scope, uri):
        """Uri activated event"""
        handled = Unity.ActivationResponse(handled=Unity.HandledType.HIDE_DASH, goto_uri=uri)
        if uri in ('vkvideo', 'vksettings'):
            subprocess.Popen([uri])
        elif self.vk:
            uri = self._last_result[int(uri)]['player']
            self.action_idle.open(uri,
                self.settings.get_string('/apps/unity-vkvideo-lens/quality') or '720',
                self.settings.get_string('/apps/unity-vkvideo-lens/player') or 'totem',
            )
        return handled

    def on_preview_uri(self, scope, uri):
        description = self._last_result[int(uri)]['description']
        uri = self._last_result[int(uri)]['player']
        page = VideoPage(uri)
        preview = Unity.GenericPreview.new(page.title, description, None)
        preview.props.image_source_uri = page.thumb
        for num, (quality, url) in enumerate(page.urls):
            action = Unity.PreviewAction.new(quality, quality + 'p', None)
            action.connect('activated', partial(self.from_preview, url, -1,
                self.settings.get_string('/apps/unity-vkvideo-lens/player') or 'totem',
            ))
            preview.add_action(action)
        return preview

    def from_preview(self, *args):
        handled = Unity.ActivationResponse(handled=Unity.HandledType.HIDE_DASH)
        self.action_idle.open(*args)
        return handled


def main():
    session_bus_connection = Gio.bus_get_sync(Gio.BusType.SESSION, None)
    session_bus = Gio.DBusProxy.new_sync(session_bus_connection, 0, None,
        'org.freedesktop.DBus',
        '/org/freedesktop/DBus',
        'org.freedesktop.DBus', None)
    result = session_bus.call_sync('RequestName',
        GLib.Variant("(su)", ('net.launchpad.Unity.Lens.VKvideo', 0x4)),
        0, -1, None)
    result = result.unpack()[0]
    if result != 1:
        raise SystemExit(1)
    daemon = Daemon(session_bus_connection)
    GObject.MainLoop().run()

if __name__ == "__main__":
    main()
